File: src/algorithms/cpp/merge-sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void emitCompare(int i, int j) {
    cout << "{\"type\":\"ARRAY_COMPARE\",\"indices\":[" << i << "," << j << "]}" << endl;
}
void emitSet(int index, int value) {
    cout << "{\"type\":\"ARRAY_SET\",\"index\":" << index << ",\"value\":" << value << "}" << endl;
}
void emitInit(const vector<int>& arr) {
    cout << "{\"type\":\"INIT\",\"array\":[";
    for (int i = 0; i < (int)arr.size(); i++) {
        if (i > 0) cout << ",";
        cout << arr[i];
    }
    cout << "]}" << endl;
}

void merge(vector<int>& arr, int l, int m, int r) {
  int n1 = m - l + 1, n2 = r - m;
  vector<int> left(arr.begin() + l, arr.begin() + m + 1);
  vector<int> right(arr.begin() + m + 1, arr.begin() + r + 1);
  int i = 0, j = 0, k = l;
  while (i < n1 && j < n2) {
    emitCompare(l + i, m + 1 + j);
    if (left[i] <= right[j]) {
      emitSet(k, left[i]);
      arr[k++] = left[i++];
    } else {
      emitSet(k, right[j]);
      arr[k++] = right[j++];
    }
  }
  while (i < n1) {
    emitSet(k, left[i]);
    arr[k++] = left[i++];
  }
  while (j < n2) {
    emitSet(k, right[j]);
    arr[k++] = right[j++];
  }
}

void mergeSort(vector<int>& arr, int left, int right) {
  if (left < right) {
    int mid = left + (right - left) / 2;
    mergeSort(arr, left, mid);
    mergeSort(arr, mid + 1, right);
    merge(arr, left, mid, right);
  }
}

int main() {
    vector<int> arr = {38, 27, 43, 3, 9, 82, 10};
    emitInit(arr);
    mergeSort(arr, 0, arr.size() - 1);
    return 0;
}

File: src/algorithms/python/merge-sort.py
import json


def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))


def merge(arr, l, m, r):
    left = arr[l : m + 1]
    right = arr[m + 1 : r + 1]
    i = j = 0
    k = l
    while i < len(left) and j < len(right):
        emit("ARRAY_COMPARE", indices=[l + i, m + 1 + j])
        if left[i] <= right[j]:
            emit("ARRAY_SET", index=k, value=left[i])
            arr[k] = left[i]
            i += 1
        else:
            emit("ARRAY_SET", index=k, value=right[j])
            arr[k] = right[j]
            j += 1
        k += 1
    while i < len(left):
        emit("ARRAY_SET", index=k, value=left[i])
        arr[k] = left[i]
        i += 1
        k += 1
    while j < len(right):
        emit("ARRAY_SET", index=k, value=right[j])
        arr[k] = right[j]
        j += 1
        k += 1


def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)


if __name__ == "__main__":
    arr = [38, 27, 43, 3, 9, 82, 10]
    emit("INIT", array=arr[:])
    merge_sort(arr, 0, len(arr) - 1)
